src/channeldef.ts

```diff
@@ -197,6 +198,154 @@ export type TextDef<F extends Field> =
   | ValueDefWithCondition<StringFieldDef<F>, Text>;
 
+/**
+ * Position of a label relative to its base mark's position, which includes anchor direction (anchor) and distance (offset).
+ */
+export type LabelPosition = {
+  /**
+   * Distance of a label from its base mark.
+   * If negative, the label position is insize.
+   */
+  offset: number;
+
+  /**
+   * Anchor direction of a label from its base mark.
+   * The value could be one of `"left"`, `"right"`, `"top"`, `"middle",`
+   * `"bottom"`, `"top-left"`, `"top-right"`, `"bottom-left"`, or "`bottom-right"`.
+   */
+  anchor: LabelAnchor;
+};
+
+type LabelMarkDef = Omit<MarkDef<'text'>, 'type'>;
+
+type AncestorLevel = {
+  /**
+   * A number `n` means `n`th ancestor layer.
+   */
+  ancestor: number;
+};
+
+type LabelAvoid = 'all-marks' | 'base-mark' | AncestorLevel;
+
+export function getAncestorLevel(avoid: LabelAvoid): number {
+  if (!avoid || avoid === 'base-mark') {
+    return 0;
+  } else if (avoid === 'all-marks') {
+    return Infinity;
+  } else {
+    return avoid.ancestor;
+  }
+}
+
+export type LabelDefMixins = {
+  /**
+   * A list of possible positions for a label to be placed relative to its base mark.
+   * The default value depends on the mark type.
+   *
+   * __Default value:__
+   * - `bar`:
+   *   - If vertical bar chart: `[
+   *       {anchor: "top", offset: 2},
+   *       {anchor: "top", offset: -2}
+   *     ]`
+   *   - If horizontal bar chart: `[
+   *       {anchor: "right", offset: 2},
+   *       {anchor: "right", offset: -2}
+   *     ]`
+   * - `line` or `trail`:
+   *   - If vertical line/trail chart:
+   *     - If `lineAnchor` is `"begin"`: `[
+   *         {anchor: "bottom-left", offset: 2},
+   *         {anchor: "bottom", offset: 2},
+   *         {anchor: "bottom-right", offset: 2},
+   *       ]`
+   *     - If `lineAnchor` is `"end"`: `[
+   *         {anchor: "top-left", offset: 2},
+   *         {anchor: "top", offset: 2},
+   *         {anchor: "top-right", offset: 2},
+   *       ]`
+   *   - If horizontal line/trail chart:
+   *     - If `lineAnchor` is `"begin"`: `[
+   *         {anchor: "top-left", offset: 2},
+   *         {anchor: "left", offset: 2},
+   *         {anchor: "bottom-left", offset: 2},
+   *       ]`
+   *     - If `lineAnchor` is `"end"`: `[
+   *         {anchor: "top-right", offset: 2},
+   *         {anchor: "right", offset: 2},
+   *         {anchor: "bottom-right", offset: 2},
+   *       ]`
+   * - `rect`: `[{anchor: "middle", offset: 0}]`
+   * - `circle` or `point` or `square`: `[
+   *     {anchor: "top-right", offset: 2},
+   *     {anchor: "top", offset: 2},
+   *     {anchor: "top-left", offset: 2},
+   *     {anchor: "left", offset: 2},
+   *     {anchor: "bottom-left", offset: 2},
+   *     {anchor: "bottom", offset: 2},
+   *     {anchor: "bottom-right", offset: 2},
+   *     {anchor: "right", offset: 2}
+   *   ]`
+   */
+  position?: LabelPosition[];
```

Comment:
Do we have label rotation?   If not, we should have it. 

for vertical bar, I'd argue that the default text should be rotated by 90 (or -90) degree because the label will better scale for arbitrary length of text -- see Tableau's label for example.

Reply:
We do not have that in Vega, right now.
Do you think it is ok to not have it right now, and we can add it later after we implement label rotation in Vega?

Reply:
> Do you think it is ok to not have it right now?

I think it's pretty bad if we can't rotate the text for vertical bars, but we also have to account for how hard it is implement there, and whether you have time for it. 

(If you don't have time, then please at least file an issue in Vega, and give some hints for how could someone else go implement the feature.)